File: api/src/main/kotlin/org/team619/rutor/config/ApiBuilder.java
package org.team619.rutor.config;

import org.jsoup.nodes.Document;
import org.team619.rutor.converter.*;
import org.team619.rutor.core.Converter;
import org.team619.rutor.core.DefaultEntity;
import org.team619.rutor.core.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ApiBuilder {

    private String baseUri;
    private Logger logger;

    private ApiBuilder(String uri) {
        this.baseUri = uri;
    }

    public static ApiBuilder from(String uri) {
        return new ApiBuilder(uri);
    }

    public ApiBuilder withLogger(Logger logger) {
        this.logger = logger;
        return this;
    }

    public RutorConverter build() {
        Objects.requireNonNull(logger, "Logger can't be null");
        List<Converter<? extends DefaultEntity, Document>> converters = new ArrayList<>();
        converters.add(new MainPagePlainConverter(logger));
        converters.add(new MainPageGroupedConverter(logger));
        converters.add(new DetailPageConverter(logger, new CardConverter(logger), new CommentConverter(logger)));
        RutorConverter rutorConverter = new RutorConverter(logger, baseUri, converters);
        return rutorConverter;
    }

}

File: api/src/main/kotlin/org/team619/rutor/config/ApiBuilder.kt
package org.team619.rutor.config

import org.jsoup.nodes.Document
import org.team619.rutor.converter.*
import org.team619.rutor.core.Converter
import org.team619.rutor.core.Logger
import org.team619.rutor.core.Page
import org.team619.rutor.core.Serializer
import java.util.*

class ApiBuilder private constructor(private val baseUri: String) {

    private var logger: Logger? = null
    private var serializer: Serializer? = null

    fun withLogger(logger: Logger): ApiBuilder {
        this.logger = logger
        return this
    }

    fun withSerializer(serializer: Serializer): ApiBuilder {
        this.serializer = serializer
        return this
    }

    fun build(): RutorConverter {
        Objects.requireNonNull<Logger>(logger, "Logger can't be null")
        Objects.requireNonNull<Serializer>(serializer, "Serializer can't be null")
        val localLogger = logger!!
        val localSerializer = serializer!!
        val cardConverter = CardConverter(localLogger)
        val commentConverter = CommentConverter(localLogger)
        val converters = ArrayList<Converter<out Page, Document>>()
        converters.add(MainPagePlainConverter(localLogger))
        converters.add(MainPageGroupedConverter(localLogger))
        converters.add(DetailPageConverter(localLogger, cardConverter, commentConverter))

        return RutorConverter(localLogger, localSerializer, baseUri, converters)
    }

    companion object {

        fun from(uri: String): ApiBuilder {
            return ApiBuilder(uri)
        }

    }

}
